homeassistant/components/firmata/config_flow.py

```diff
@@ -0,0 +1,99 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class FirmataFlowHandler(config_entries.ConfigFlow):
+    """Handle a firmata config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    _hassio_discovery = None
+
+    def __init__(self):
+        """Initialize the firmata config flow."""
+        self.firmata_config = {}
+
+    async def async_step_init(self, user_input=None):
+        """Needed in order to not require re-translation of strings."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a firmata config flow start.
+
+        If only one board is found go to link step.
+        If more than one board is found let user choose board to link.
+        If no board is found allow user to manually input configuration.
+        """
+        if user_input is not None:
+            self.firmata_config = user_input
+            return await self._create_entry()
+
+        return self.async_show_form(
+            step_id='init',
+            data_schema=vol.Schema({
+                vol.Required(CONF_SERIAL_PORT): str
+            }),
+        )
+
+#    @staticmethod
+#    def serial_ports():
+#        """ List possible serial ports.
+#        From https://stackoverflow.com/a/14224477"""
+#        import glob
+#        import sys
+#        if sys.platform.startswith('win'):
+#            ports = ['COM%s' % (i + 1) for i in range(256)]
+#        elif (sys.platform.startswith('linux') or
+#              sys.platform.startswith('cygwin')):
+#            # this excludes your current terminal "/dev/tty"
+#            ports = glob.glob('/dev/tty[A-Za-z]*')
+#        elif sys.platform.startswith('darwin'):
+#            ports = glob.glob('/dev/tty.*')
+#        return ports
+
+    async def _create_entry(self):
+        """Create entry for board."""
+        if CONF_SERIAL_PORT in self.firmata_config:
+            name = self.firmata_config[CONF_SERIAL_PORT]
+            if '/' in name:
+                name = name.split('/')[-1]
+        elif CONF_PORT in self.firmata_config:
+            name = (self.firmata_config[CONF_HOST] + '-' +
+                    self.firmata_config[CONF_PORT])
+        else:
+            name = self.firmata_config[CONF_HOST]
+        self.firmata_config[CONF_NAME] = name
+
+        return self.async_create_entry(
+            title='firmata-' + self.firmata_config[CONF_NAME],
+            data=self.firmata_config
+        )
+
+    async def async_step_import(self, import_config):
+        """Import a firmata board as a config entry.
+
+        This flow is triggered by `async_setup` for configured boards.
+        This flow is also triggered by `async_step_discovery`.
+
+        This will execute for any board that does not have a
+        config entry yet (based on board name).
+        """
+        self.firmata_config = import_config
+
+        return await self._create_entry()
```

Comment:
I'm confused where I should actually make a connection to the board. Should I move initializing PyMata to `async_step_import`? Or should I make a connection twice (this isn't ideal because it can take up to ten seconds)? I'll move it for the next commit. Let me know if I should do something else.

Reply:
Wait will `async_step_import` not be called for existing entries? It definitely calls `async_setup_entry` every time.

Reply:
We should check that we don't start the config flow, via import step, for existing config sections that match existing config entries.

Reply:
Sorry, but I'm still confused what I should do. If the flow isn't started, then won't `async_setup_entry` not run and the board not be setup? And would I not want to overwrite the config entries every time so that changes from `configuration.yaml` propagate to the entries?

Reply:
It's common to pass the config from the import step to the user step. From the user step we check the connection before creating the entry.

We have to decide if a change in the configuration.yaml should remove the existing config entry and create a new entry and how to handle that. If there's no change in configuration.yaml it's not necessary to create a new entry. Maybe checksum the config and store in the entry data and compare to the checksum of the config in the import step?

Reply:
I was going to implement deleting the entries and re-adding them in `async_setup` as you recommended before:
> You could remove the existing entries that match the ones in configuration.yaml and import the updated config as new entries, automatically.

I just attempted implementing the following:
```python3
async def async_setup(hass, config):
    """Set up the Firmata domain."""
    if not hass.config_entries.async_entries(DOMAIN) and DOMAIN in config:
        for board in config[DOMAIN]:
            firmata_config = deepcopy(board)
            hass.async_create_task(hass.config_entries.flow.async_init(
                DOMAIN, context={'source': config_entries.SOURCE_IMPORT},
                data=firmata_config
            ))
    elif DOMAIN in config:
        for board in config[DOMAIN]:
            firmata_config = deepcopy(board)
            update_board = False
            for entry in hass.config_entries.async_entries(DOMAIN):
                if (CONF_SERIAL_PORT in firmata_config and
                    CONF_SERIAL_PORT in entry.data and
                    firmata_config[CONF_SERIAL_PORT] ==
                    entry.data[CONF_SERIAL_PORT]):
                    update_board = entry
                    break
                elif (CONF_REMOTE in firmata_config and
                      CONF_REMOTE in entry.data and
                      CONF_PORT in firmata_config[CONF_REMOTE] and
                      CONF_PORT in entry.data[CONF_REMOTE] and
                      firmata_config[CONF_REMOTE][CONF_HOST] ==
                      entry.data[CONF_REMOTE][CONF_HOST] and
                      firmata_config[CONF_REMOTE][CONF_PORT] ==
                      entry.data[CONF_REMOTE][CONF_PORT]):
                    update_board = entry
                    break
                elif (CONF_REMOTE in firmata_config and
                      CONF_REMOTE in entry.data and
                      CONF_PORT not in firmata_config[CONF_REMOTE] and
                      CONF_PORT not in entry.data[CONF_REMOTE] and
                      firmata_config[CONF_REMOTE][CONF_HOST] ==
                      entry.data[CONF_REMOTE][CONF_HOST]):
                    update_board = entry
                    break
            if update_board:
                update_board.data.update(firmata_config)
                hass.config_entries.async_update_entry(update_board,
                                                       data=update_board.data)
            else:
                hass.async_create_task(hass.config_entries.flow.async_init(
                    DOMAIN, context={'source': config_entries.SOURCE_IMPORT},
                    data=firmata_config
                ))
    return True
```
Basically, if there is already a config entry for a board (identified by having the same serial port, host & port, or host), then it will update it with the new config from `configuration.yaml`. However, this created an issue where if I modified a `switch` or `binary_sensor` (such as by changing its pin assignment), it would create a new entity for the updated item (as intended) but it would not remove the defunct item of the same name. So, updating the entity doesn't work as intended unless all of the entities keep the same identifiers. Identifying by name instead of pin would not help, as if a user changed the name of an item, the old one would still linger.

I think I understand what you are saying now:
1. In `async_setup`, add the board's config to the entry data.
2. In `async_step_import`, checksum the config from `configuration.yaml` by some method (pickle, JSON string, JSON string -> SHA1 or something).
3. In `async_step_import`, if the entity already exists, check if the checksum has changed.
   - If it changed, remove the entry and return a new entry.
   - If it is a new entry, return the new entry.
   - If it didn't change, return `False`.

How could I do this "automatically" as you mentioned before? Both of these methods seem very elaborate, and I saw nothing like this while looking through other component sources.

Reply:
Checking and replacing a config entry have to be done explicitly. It isn't handled by the core.

Existing config entries will be set up automatically, unless we remove them before that happens. We can do the check in the config flow import step. If the existing config entry doesn't need to be replaced we just abort the flow.